File: frontend/src/api/alerts.ts
import { request } from './request'

export interface AlertItem {
  id: number
  student_id: number
  student_name: string | null
  class_id: number | null
  level: 'green' | 'yellow' | 'orange' | 'red'
  score: number
  title: string
  reasons: string[]
  sources: string[]
  status: 'open' | 'acknowledged' | 'resolved' | 'closed'
  handler_id: number | null
  acknowledged_at: string | null
  resolved_at: string | null
  closed_at: string | null
  created_at: string
}

export interface AlertStats {
  total: number
  by_level: Record<string, number>
  by_status: Record<string, number>
}

export interface CorrelationItem {
  student_id: number
  name: string
  class_id: number
  psy_30d: number
  score_avg: number
  scale_score: number
  phone_count: number
  risk_score: number
  level: string
}

export const listAlerts = (params?: any) =>
  request<{ items: AlertItem[]; total: number }>({ url: '/alerts', method: 'get', params })

export const getAlertStats = () =>
  request<AlertStats>({ url: '/alerts/stats', method: 'get' })

export const recomputeAlerts = () =>
  request<{ processed: number; alerts_count: number; alerts: AlertItem[] }>({
    url: '/alerts/recompute',
    method: 'post',
  })

export const ackAlert = (id: number) =>
  request<AlertItem>({ url: `/alerts/${id}/acknowledge`, method: 'post' })

export const resolveAlert = (id: number, note?: string) =>
  request<AlertItem>({ url: `/alerts/${id}/resolve`, method: 'post', data: { note } })

export const closeAlert = (id: number) =>
  request<AlertItem>({ url: `/alerts/${id}/close`, method: 'post' })

export const listInterventions = (alertId: number) =>
  request<{ items: any[] }>({ url: `/alerts/${alertId}/interventions`, method: 'get' })

export const addIntervention = (alertId: number, data: any) =>
  request<any>({ url: `/alerts/${alertId}/interventions`, method: 'post', data })

export const getCorrelationMatrix = () =>
  request<{ items: CorrelationItem[]; fields: { key: string; label: string }[] }>({
    url: '/alerts/correlation/matrix',
    method: 'get',
  })

File: backend/app/api/alerts.py
"""预警与关联分析接口。"""

from __future__ import annotations

from flask import Blueprint, request

from ..services import correlation_service as svc
from ..utils.permissions import (
    get_current_user,
    get_visible_student_ids,
    login_required,
)
from ..utils.response import ok

alerts_bp = Blueprint("alerts", __name__)


@alerts_bp.get("")
@login_required
def list_alerts():
    user = get_current_user()
    visible = get_visible_student_ids(user)
    return ok(
        svc.list_alerts(
            level=request.args.get("level"),
            status=request.args.get("status"),
            student_ids=visible,
            page=request.args.get("page", default=1, type=int),
            page_size=request.args.get("page_size", default=20, type=int),
        )
    )


@alerts_bp.get("/stats")
@login_required
def alert_stats():
    user = get_current_user()
    visible = get_visible_student_ids(user)
    return ok(svc.alert_stats(student_ids=visible))


@alerts_bp.post("/recompute")
@login_required
def recompute():
    user = get_current_user()
    visible = get_visible_student_ids(user)
    return ok(svc.recompute_alerts_for_visible(student_ids=visible), "已重算")


@alerts_bp.post("/<int:alert_id>/acknowledge")
@login_required
def acknowledge(alert_id: int):
    user = get_current_user()
    return ok(svc.acknowledge_alert(alert_id, user.id), "已签收")


@alerts_bp.post("/<int:alert_id>/resolve")
@login_required
def resolve(alert_id: int):
    user = get_current_user()
    payload = request.get_json() or {}
    return ok(svc.resolve_alert(alert_id, user.id, payload.get("note")), "已处理")


@alerts_bp.post("/<int:alert_id>/close")
@login_required
def close(alert_id: int):
    user = get_current_user()
    return ok(svc.close_alert(alert_id, user.id), "已关闭")


@alerts_bp.get("/<int:alert_id>/interventions")
@login_required
def list_interventions(alert_id: int):
    return ok({"items": svc.list_interventions(alert_id)})


@alerts_bp.post("/<int:alert_id>/interventions")
@login_required
def add_intervention(alert_id: int):
    user = get_current_user()
    payload = request.get_json() or {}
    return ok(
        svc.add_intervention(
            alert_id,
            user.id,
            payload.get("action", "其他"),
            payload.get("summary", ""),
            payload.get("follow_up"),
        ),
        "已记录",
    )


# 关联分析
@alerts_bp.get("/correlation/matrix")
@login_required
def correlation_matrix():
    user = get_current_user()
    visible = get_visible_student_ids(user)
    return ok(svc.correlation_matrix(visible))
